python_modules/libraries/dagster-gcp/dagster_gcp/bigquery/resources.py

```diff
@@ -1,20 +1,54 @@
-from dagster import resource
+from contextlib import contextmanager
+from typing import Optional
+
+from dagster import ConfigurableResource, resource
 from google.cloud import bigquery
+from pydantic import Field
 
-from .configs import bq_resource_config
 from .utils import setup_gcp_creds
 
 
+class BigQueryResource(ConfigurableResource):
+    project: Optional[str] = Field(
+        default=None,
+        description="""Project ID for the project which the client acts on behalf of. Will be passed
+        when creating a dataset / job. If not passed, falls back to the default inferred from the
+        environment.""",
+    )
+
+    location: Optional[str] = Field(
+        default=None,
+        description="Default location for jobs / datasets / tables.",
+    )
+
+    gcp_credentials: Optional[str] = Field(
+        default=None,
+        description=(
+            "GCP authentication credentials. If provided, a temporary file will be created"
+            " with the credentials and GOOGLE_APPLICATION_CREDENTIALS will be set to the"
+            " temporary file. To avoid issues with newlines in the keys, you must base64"
+            " encode the key. You can retrieve the base64 encoded key with this shell"
+            " command: cat $GOOGLE_AUTH_CREDENTIALS | base64"
+        ),
+    )
+
+    @contextmanager
+    def setup_credentials(self):
+        if self.gcp_credentials:
+            with setup_gcp_creds(self.gcp_credentials):
+                yield
+        else:
+            yield
+
+    def get_client(self) -> bigquery.Client:
+        return bigquery.Client(project=self.project, location=self.location)
+
```

Comment:
Stepping back for a second, why do we need this pattern in the first place? It seems pretty contorted to have to provided a base64-encoded key and write it out to a temp file like this.

Reply:
Assuming we need this, IMO we should have the `get_client` be a context manager. Like so:

```
with bq_resource.client() as client:
   client.whatever()
```

Prevents class of programming errors as right now you could construct a client in a way that will break in non-obvious ways if you forgot to call `setup_credentials`

Reply:
mainly it's to support serverless. to auth with gcp you either need to run `gcloud auth login` or set an env var `GOOGLE_APPLICATION_CREDENTIALS=creds_file.json` where `creds_file.json` has your auth key. For serverless, users can't upload a creds file, or run the interactive `gcloud auth login` command so we do this sort of work around to allow them to pass the raw credentials string and then we write it to a file and set the env var. 

The base64 encoding is to solve an issue where newlines in the creds file will cause authentication to fail. by base64 encoding it, we dont have to worry about newlines

as for having `get_client` be a context manager - are you thinking like this instead?
```
@contextmanager
    def get_client(self):
        if self.gcp_credentials:
            with setup_gcp_creds(self.gcp_credentials):
                yield bigquery.Client(project=self.project, location=self.location)
        else:
            yield bigquery.Client(project=self.project, location=self.location)

@asset 
def my_asset(bigquery: BigQueryResource):
    with bigquery.get_client() as client:
        client.query(....)
```

Reply:
Yeah that's what I'm thinking. It unfortunately imposes some overhead on people who wouldn't have any use for the `gcp_credentials` feature with a strictly unnecessary `with` block.

I think having a `setup_credentials` is a bit odd since if you have the env var set it does work. `setup_credentials` sounds like something you are required to call in all circumstances.

Reply:
yeah i think moving to `get_client` being the context manager makes sense. I was playing around with a duckdb resource and i think the `get_client` method there will also need to be a context manager so we have handle closing the connection. This way everything is consistent with `get_client` being a context manager for some resources